### PR TITLE
[SPARK-9482] [SQL] Fix thread-safey issue of using UnsafeProjection in join

### PR DESCRIPTION
**Pull request by davies**
This PR also change to use `def` instead of `lazy val` for UnsafeProjection, because it's not thread safe.

TODO: cleanup the debug code once the flaky test passed 100 times.

### CODE REVIEW

**Comment by rxin** [comment]
@davies it might be worth it to have a standalone unit test to runs some broadcast join query 100 times concurrently.

**Comment by davies** [comment]
Once we fix this, I think it's not that necessary, or it just slow down the whole tests.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala [hunk 1]
```diff
@@ -59,19 +57,17 @@ case class BroadcastHashJoin(
   override def requiredChildDistribution: Seq[Distribution] =
     UnspecifiedDistribution :: UnspecifiedDistribution :: Nil
 
-  @transient
-  private val broadcastFuture = future {
+  private lazy val broadcastedRelation = {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: There was a noticeable speed-up in TPC-DS from broadcasting in parallel and I think this removes that.  Perhaps we should both use a `lazy val` and a `future`.  We can have the prepare phase make sure the future all get invoked before running.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: I see, we can hold this until #7774 merged in.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: Could you pull this out into its own PR then and fix the actual issue here?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: #7774 also touch here, it also could fix the race condition, should we wait for that?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: Looks like that has been merged.  Is it fixed now?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala]: I think so, will re-purpose this PR to fix the thead-safy issue about UnsafeProjection.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoin.scala [hunk 2]
```diff
@@ -47,7 +47,7 @@ case class BroadcastNestedLoopJoin(
   override def outputsUnsafeRows: Boolean = left.outputsUnsafeRows || right.outputsUnsafeRows
   override def canProcessUnsafeRows: Boolean = true
 
-  @transient private[this] lazy val resultProjection: InternalRow => InternalRow = {
+  @transient private[this] def genResultProjection: InternalRow => InternalRow = {
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoin.scala]: ```@transient def```? Does that work? Maybe not the most constructive feedback, but it stood out...

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoin.scala]: Good catch, @transient should be removed

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoin.scala]: How does this work? Right now we generate the class on the driver and than ship the generated closure. Shouldn't we generate the class on the executors?

**Comment by davies** [comment]
The semijoin test had ran 100 times without failure, so we could think that is fixed. I'm going to remove the debug code, the fix others.
